T6841: firewall: improve config parsing for ZBF when using VRFs and interfaces attached to VRFs

nicolas-fort opened the pull request:
<!-- All PR should follow this template to allow a clean and transparent review -->
<!-- Text placed between these delimiters is considered a comment and is not rendered -->

## Change Summary
<!--- Provide a general summary of your changes in the Title above -->
Improve config parsing for ZBF when using VRFs and interfaces attached to VRFs

## Types of changes
<!---
What types of changes does your code introduce? Put an 'x' in all the boxes that apply.
NOTE: Markdown requires no leading or trailing whitespace inside the [ ] for checking
the box, please use [x]
-->
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes)
- [ ] Migration from an old Vyatta component to vyos-1x, please link to related PR inside obsoleted component
- [ ] Other (please describe):

## Related Task(s)
<!-- optional: Link to related other tasks on Phabricator. -->
<!-- * https://vyos.dev/Txxxx -->
* https://vyos.dev/T6841

## Related PR(s)
<!-- Link here any PRs in other repositories that are required by this PR -->

## Component(s) name
<!-- A rather incomplete list of components: ethernet, wireguard, bgp, mpls, ldp, l2tp, dhcp ... -->
firewall

## Proposed changes
<!--- Describe your changes in detail -->
For zone based firewall, everything is related to matching inbound and outbound interface. The problem is that in Linux, if an interface is attached to a non-default VRF, then:
 - For matching inbound-interface, we need to specify <vrf_name> in firewall ruleset
 - For matching outbound-interface, we need to specify <physical_interface> in firewall ruleset.

Before this PR, what was written under `set firewall zone <zone> interface <iface>` was exactly written for inbound|outbound interface in nftables.
Now we have provide more options so we can specify `interface name` and `interfave vrf` while defining interfaces in a zone.
- If using `interface name <iface>` --> it still writes exactly that interfaces for inbound|outbound interface in nftables
- If using `interface vrf <vrf_name>` --> in nftables it writes:
    -  <vrf_name> for matching inbound interface
    - <all_vrf_members> for matching outbound interface.


## How to test
<!---
Please describe in detail how you tested your changes. Include details of your testing
environment, and the tests you ran. When pasting configs, logs, shell output, backtraces,
and other large chunks of text, surround this text with triple backtics
```
like this
```
-->

## Smoketest result
<!-- Provide the output of the smoketest
```
$ /usr/libexec/vyos/tests/smoke/cli/test_xxx_feature.py
test_01_simple_options (__main__.TestFeature.test_01_simple_options) ... ok
```
-->
```
root@zbf-vrfaware:/usr/libexec/vyos/tests/smoke/cli# ./test_firewall.py 
test_bridge_firewall (__main__.TestFirewall.test_bridge_firewall) ... ok
test_cyclic_jump_validation (__main__.TestFirewall.test_cyclic_jump_validation) ... ok
test_flow_offload (__main__.TestFirewall.test_flow_offload) ... ok
test_geoip (__main__.TestFirewall.test_geoip) ... ok
test_gre_match (__main__.TestFirewall.test_gre_match) ... ok
test_groups (__main__.TestFirewall.test_groups) ... ok
test_ipsec_metadata_match (__main__.TestFirewall.test_ipsec_metadata_match) ... ok
test_ipv4_advanced (__main__.TestFirewall.test_ipv4_advanced) ... ok
test_ipv4_basic_rules (__main__.TestFirewall.test_ipv4_basic_rules) ... ok
test_ipv4_dynamic_groups (__main__.TestFirewall.test_ipv4_dynamic_groups) ... ok
test_ipv4_global_state (__main__.TestFirewall.test_ipv4_global_state) ... ok
test_ipv4_mask (__main__.TestFirewall.test_ipv4_mask) ... ok
test_ipv4_state_and_status_rules (__main__.TestFirewall.test_ipv4_state_and_status_rules) ... ok
test_ipv4_synproxy (__main__.TestFirewall.test_ipv4_synproxy) ... ok
test_ipv6_advanced (__main__.TestFirewall.test_ipv6_advanced) ... ok
test_ipv6_basic_rules (__main__.TestFirewall.test_ipv6_basic_rules) ... ok
test_ipv6_dynamic_groups (__main__.TestFirewall.test_ipv6_dynamic_groups) ... ok
test_ipv6_mask (__main__.TestFirewall.test_ipv6_mask) ... ok
test_nested_groups (__main__.TestFirewall.test_nested_groups) ... ok
test_source_validation (__main__.TestFirewall.test_source_validation) ... ok
test_sysfs (__main__.TestFirewall.test_sysfs) ... ok
test_timeout_sysctl (__main__.TestFirewall.test_timeout_sysctl) ... ok
test_zone_basic (__main__.TestFirewall.test_zone_basic) ... ok
test_zone_flow_offload (__main__.TestFirewall.test_zone_flow_offload) ... ok
test_zone_with_vrf (__main__.TestFirewall.test_zone_with_vrf) ... ok

----------------------------------------------------------------------
Ran 25 tests in 136.776s

OK

```


## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
<!--- The entire development process is outlined here: https://docs.vyos.io/en/latest/contributing/development.html -->
- [x] I have read the [**CONTRIBUTING**](https://github.com/vyos/vyos-1x/blob/current/CONTRIBUTING.md) document
- [x] I have linked this PR to one or more Phabricator Task(s)
- [x] I have run the components [**SMOKETESTS**](https://github.com/vyos/vyos-1x/tree/current/smoketest/scripts/cli) if applicable
- [x] My commit headlines contain a valid Task id
- [x] My change requires a change to the documentation
- [ ] I have updated the documentation accordingly

nicolas-fort committed:
T6841: firewall: improve config parsing for ZBF when using VRFs and interfaces attached to VRFs

nicolas-fort commented:
Should I also change these files?
https://github.com/search?q=repo%3Avyos%2Fvyos-1x+path%3A%2F%5Esmoketest%5C%2Fconfig-tests%5C%2F%2F+%22set+firewall+zone%22&type=code

dmbaturin commented on interface-definitions/firewall.xml.in:
```diff
@@ -464,24 +464,39 @@
             </children>
           </tagNode>
-          <leafNode name="interface">
+          <node name="interface">
```
```suggestion
          <node name="member">
```

I agree with the idea to separate interface and VRF options but I have reservations about this syntax. The problem is that a VRF is not an interface, any VRF may include as many interfaces as it wants.

So the parent node should be named something else than `interface`. I think `member` might be a good generic word.

dmbaturin commented on interface-definitions/firewall.xml.in:
```diff
@@ -464,24 +464,39 @@
             </properties>
-          </leafNode>
+            <children>
+              <leafNode name="name">
```
```suggestion
              <leafNode name="interface">
```

dmbaturin commented on smoketest/scripts/cli/test_firewall.py:
```diff
@@ -905,7 +905,7 @@ def test_timeout_sysctl(self):
         self.cli_set(['firewall', 'ipv4', 'name', 'smoketest', 'default-action', 'drop'])
         self.cli_set(['firewall', 'ipv6', 'name', 'smoketestv6', 'default-action', 'drop'])
-        self.cli_set(['firewall', 'zone', 'smoketest-eth0', 'interface', 'eth0'])
+        self.cli_set(['firewall', 'zone', 'smoketest-eth0', 'interface', 'name', 'eth0'])
```
```suggestion
        self.cli_set(['firewall', 'zone', 'smoketest-eth0', 'member', 'interface', 'eth0'])
```

dmbaturin commented on smoketest/scripts/cli/test_firewall.py:
```diff
@@ -963,6 +963,98 @@ def test_zone_basic(self):
+        self.cli_set(['firewall', 'zone', 'LOCAL', 'from', 'ZONE1', 'firewall', 'name', 'ZONE1-to-LOCAL'])
+        self.cli_set(['firewall', 'zone', 'LOCAL', 'local-zone'])
+        self.cli_set(['firewall', 'zone', 'ZONE1', 'from', 'ZONE2', 'firewall', 'name', 'ZONE2_to_ZONE1'])
+        self.cli_set(['firewall', 'zone', 'ZONE1', 'interface', 'name', 'eth1'])
```
```suggestion
        self.cli_set(['firewall', 'zone', 'ZONE1', 'member', 'interface', 'eth1'])
```

dmbaturin commented on smoketest/scripts/cli/test_firewall.py:
```diff
@@ -963,6 +963,98 @@ def test_zone_basic(self):
+        self.cli_set(['firewall', 'zone', 'LOCAL', 'local-zone'])
+        self.cli_set(['firewall', 'zone', 'ZONE1', 'from', 'ZONE2', 'firewall', 'name', 'ZONE2_to_ZONE1'])
+        self.cli_set(['firewall', 'zone', 'ZONE1', 'interface', 'name', 'eth1'])
+        self.cli_set(['firewall', 'zone', 'ZONE1', 'interface', 'name', 'eth2'])
```
```suggestion
        self.cli_set(['firewall', 'zone', 'ZONE1', 'member', 'interface', 'eth2'])
```

dmbaturin commented on smoketest/scripts/cli/test_firewall.py:
```diff
@@ -963,6 +963,98 @@ def test_zone_basic(self):
+        self.cli_set(['firewall', 'zone', 'ZONE1', 'from', 'ZONE2', 'firewall', 'name', 'ZONE2_to_ZONE1'])
+        self.cli_set(['firewall', 'zone', 'ZONE1', 'interface', 'name', 'eth1'])
+        self.cli_set(['firewall', 'zone', 'ZONE1', 'interface', 'name', 'eth2'])
+        self.cli_set(['firewall', 'zone', 'ZONE1', 'interface', 'vrf', 'VRF-1'])
```
```suggestion
        self.cli_set(['firewall', 'zone', 'ZONE1', 'member', 'vrf', 'VRF-1'])
```

dmbaturin commented on smoketest/scripts/cli/test_firewall.py:
```diff
@@ -963,6 +963,98 @@ def test_zone_basic(self):
+        self.cli_set(['firewall', 'zone', 'ZONE1', 'interface', 'name', 'eth2'])
+        self.cli_set(['firewall', 'zone', 'ZONE1', 'interface', 'vrf', 'VRF-1'])
+        self.cli_set(['firewall', 'zone', 'ZONE2', 'from', 'LOCAL', 'firewall', 'ipv6-name', 'LOCAL_to_ZONE2_v6'])
+        self.cli_set(['firewall', 'zone', 'ZONE2', 'interface', 'name', 'vtun66'])
```
```suggestion
        self.cli_set(['firewall', 'zone', 'ZONE2', 'member', 'interface', 'vtun66'])
```

dmbaturin commented on smoketest/scripts/cli/test_firewall.py:
```diff
@@ -963,6 +963,98 @@ def test_zone_basic(self):
+        self.cli_set(['firewall', 'zone', 'ZONE1', 'interface', 'vrf', 'VRF-1'])
+        self.cli_set(['firewall', 'zone', 'ZONE2', 'from', 'LOCAL', 'firewall', 'ipv6-name', 'LOCAL_to_ZONE2_v6'])
+        self.cli_set(['firewall', 'zone', 'ZONE2', 'interface', 'name', 'vtun66'])
+        self.cli_set(['firewall', 'zone', 'ZONE2', 'interface', 'vrf', 'VRF-2'])
```
```suggestion
        self.cli_set(['firewall', 'zone', 'ZONE2', 'member', 'vrf', 'VRF-2'])
```

dmbaturin commented on src/conf_mode/firewall.py:
```diff
@@ -458,12 +461,23 @@ def verify(firewall):
 
             if 'interface' in zone_conf:
-                found_duplicates = [intf for intf in zone_conf['interface'] if intf in zone_interfaces]
+                if 'name'in zone_conf['interface']:
```
```suggestion
                if 'interface' in zone_conf['member']:
```

dmbaturin commented on src/conf_mode/firewall.py:
```diff
@@ -458,12 +461,23 @@ def verify(firewall):
 
-                if found_duplicates:
-                    raise ConfigError(f'Interfaces cannot be assigned to multiple zones')
+                    for iface in zone_conf['interface']['name']:
```
```suggestion
                    for iface in zone_conf['member']['interface']:
```

dmbaturin commented on src/conf_mode/firewall.py:
```diff
@@ -458,12 +461,23 @@ def verify(firewall):
+                            Warning(f"Interface {iface} assigned to zone {zone} is in VRF {iface_vrf}. This might not work as expected.")
+                        zone_interfaces += iface
+
+                if 'vrf' in zone_conf['interface']:
```
```suggestion
                if 'vrf' in zone_conf['member']:
```

dmbaturin commented on src/conf_mode/firewall.py:
```diff
@@ -458,12 +461,23 @@ def verify(firewall):
+                        zone_interfaces += iface
+
+                if 'vrf' in zone_conf['interface']:
+                    for vrf in zone_conf['interface']['vrf']:
```
```suggestion
                    for vrf in zone_conf['member']['vrf']:
```

dmbaturin commented on src/conf_mode/firewall.py:
```diff
@@ -505,6 +519,12 @@ def generate(firewall):
         for local_zone, local_zone_conf in firewall['zone'].items():
             if 'local_zone' not in local_zone_conf:
+                # Get physical interfaces assigned to the zone if vrf is used:
+                if 'vrf' in local_zone_conf['interface']:
```
```suggestion
                if 'vrf' in local_zone_conf['member']:
```

dmbaturin commented on src/conf_mode/firewall.py:
```diff
@@ -505,6 +519,12 @@ def generate(firewall):
+                # Get physical interfaces assigned to the zone if vrf is used:
+                if 'vrf' in local_zone_conf['interface']:
+                    local_zone_conf['vrf_interfaces'] = {}
+                    for vrf_name in local_zone_conf['interface']['vrf']:
```
```suggestion
                    for vrf_name in local_zone_conf['member']['vrf']:
```

dmbaturin commented on src/migration-scripts/firewall/17-to-18:
```diff
@@ -0,0 +1,36 @@
+# From
+    # set firewall zone <zone> interface <iface>
+# To
+    # set firewall zone <zone> interface name <iface>
```
```suggestion
    # set firewall zone <zone> member interface <iface>
```

dmbaturin commented on src/migration-scripts/firewall/17-to-18:
```diff
@@ -0,0 +1,36 @@
+# To
+    # set firewall zone <zone> interface name <iface>
+    # or
+    # set firewall zone <zone> interface vrf <vrf>
```
```suggestion
    # set firewall zone <zone> member vrf <vrf>
```

dmbaturin commented on src/migration-scripts/firewall/17-to-18:
```diff
@@ -0,0 +1,36 @@
+    for zone in config.list_nodes(base):
+        if config.exists(base + [zone, 'interface']):
+            for iface in config.return_values(base + [zone, 'interface']):
+                config.set(base + [zone, 'interface', 'name'], value=iface, replace=False)
```
```suggestion
                config.set(base + [zone, 'member', 'interface'], value=iface, replace=False)
```

dmbaturin commented on src/migration-scripts/firewall/17-to-18:
```diff
@@ -0,0 +1,36 @@
+        # Nothing to do
+        return
+
+    for zone in config.list_nodes(base):
```
This logic looks incomplete.

1. It doesn't seem to delete values from under the old `interface` node, only duplicates them under the new one.
2. VRF support was introduced in 1.4.0 already, but the script doesn't handle the case when the value is a VRF name rather than an interface name.

c-po replied:
Addressed your comment in the latest commit. Tested by @andriiandrieiev

dmbaturin reviewed (changes requested):
I suggest a different syntax that I think is more intuitive (or less counter-intuitive ;) and there are some missing cases in the migration script logic.

aapostoliuk committed:
T6841: firewall: Fixed issues in ZBF when using VRFs

Improve config parsing for ZBF when using VRFs and interfaces attached to VRFs

c-po commented on interface-definitions/firewall.xml.in:
```diff
@@ -464,24 +464,39 @@
             </properties>
-          </leafNode>
+            <children>
+              <leafNode name="interface">
```
Please reference the generic interface include https://github.com/vyos/vyos-1x/blob/current/interface-definitions/include/generic-interface-multi.xml.i

c-po committed:
T6841: firewall: re-use existing generic-interface-multi.xml.i XML building block

c-po committed:
T6841: firewall: migrate existing VRF in zone based firewall

VRF support was introduced in VyOS 1.4.0. If a VRF is added as an interface in
the zone based firewall, it will be migrated to the new syntax.

OLD:
  set firewall zone FOO interface RED
  set firewall zone FOO interface eth0

NEW:
  set firewall zone FOO member vrf RED
  set firewall zone FOO member interface eth0
